drivers/flash/flash_util.c

```diff
@@ -80,3 +87,199 @@ int z_impl_flash_flatten(const struct device *dev, off_t offset, size_t size)
 	return -ENOSYS;
 #endif
 }
+
+static bool off_add_overflow(off_t offset, off_t size, off_t *result)
+{
+	if (sizeof(off_t) == sizeof(uint32_t)) {
+		uint32_t end;
+
+		/* account for signedness of off_t due to lack of s32_add_overflow() */
+		if (u32_add_overflow((uint32_t)offset, (uint32_t)size, &end) ||
+		    (end > (UINT32_MAX >> 1))) {
+			return true;
+		}
+	} else if (sizeof(off_t) == sizeof(uint64_t)) {
+		uint64_t end;
+
+		/* account for signedness of off_t due to lack of s64_add_overflow() */
+		if (u64_add_overflow((uint64_t)offset, (uint64_t)size, &end) ||
+		    (end > (UINT64_MAX >> 1))) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+int flash_regions_overlap(off_t a_start, off_t a_size, off_t b_start, off_t b_size)
+{
+	off_t a_end = 0;
+	off_t b_end = 0;
+
+	BUILD_ASSERT((sizeof(off_t) == sizeof(uint32_t)) || (sizeof(off_t) == sizeof(uint64_t)));
+
+	if ((a_start < 0) || (a_size < 0) || (b_start < 0) || (b_size < 0)) {
+		return -EINVAL;
+	}
+
+	if (off_add_overflow(a_start, a_size, &a_end) ||
+	    off_add_overflow(b_start, b_size, &b_end)) {
+		return -EOVERFLOW;
+	}
+
+	return (a_start < b_end) && (a_end > b_start);
+}
+
+int flash_region_crc(const struct device *dev, off_t offset, off_t size, uint32_t *crc,
+		     uint32_t *buf, size_t buf_size)
+{
+	off_t end;
+	uint32_t _buf[4];
+
+	if (!IS_ENABLED(CONFIG_CRC)) {
+		return -ENOTSUP;
+	}
+
+	if ((crc == NULL) || (offset < 0) || (size < 0)) {
+		return -EINVAL;
+	}
+
+	if (off_add_overflow(offset, size, &end)) {
+		return -EOVERFLOW;
+	}
+
+	if ((buf == NULL) || (buf_size == 0)) {
+		buf = _buf;
+		buf_size = sizeof(_buf);
+	}
+
+	if ((buf_size & (sizeof(uint32_t) - 1)) != 0) {
+		LOG_DBG("buffer size is not a multiple of uint32_t size");
+		return -EINVAL;
+	}
+
+	if (!device_is_ready(dev)) {
+		LOG_DBG("device not ready");
+		return -ENODEV;
+	}
+
+	for (uint32_t bytes = 0; size > 0; offset += bytes, size -= bytes) {
+		bytes = MIN(size, buf_size);
+		if (flash_read(dev, offset, buf, bytes) < 0) {
+			return -EIO;
+		}
+
+		*crc = crc32_ieee_update(*crc, (uint8_t *)buf, bytes);
+	}
+
+	return 0;
+}
+
+int flash_copy(const struct device *src_dev, off_t src_offset, const struct device *dst_dev,
+	       off_t dst_offset, off_t size, bool verify, uint32_t *buf, size_t buf_size)
+{
+	int ret;
+	uint32_t _buf[4];
```

Comment:
I wonder if it's possible for it to be done in Devicetree macros.

Something compile-time constant would be ideal. Of course it would only apply to the devices known at build time.

Reply:
The problem is that you can have drivers that do not even touch DTS, so you have to have an option to  override that in Kconfig too.
I also do not like picking such things from DTS, because it basically means that the value leaks through several layers, and ends up having unforeseen consequences.

Reply:
Fixed (caller should erase prior to copy)